dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/OpenTelemetryInfoFactory.java

```diff
@@ -0,0 +1,358 @@
+@Component(property = { "service.vendor=IBM",  "service.ranking:Integer=150" })
+public class OpenTelemetryInfoFactory implements ApplicationStateListener {
+	
+	/*
+	 * Class summary. 
+	 * 
+	 * This class consists of three parts
+	 * 1: Static methods for producing Objects of: OpenTelemetryInfo, OpenTelemetry, Tracer, Span, Baggage.
+	 *    If invoked after an app has been shutdown these will return a no-op object.
+	 * 
+	 * 2. A cache of OpenTelemetryInfo (using J2EENames as the key). OpenTelemetryInfo in turn contains OpenTelemetry objects. This is populated lazily using computeIfAbsent and the use of ConcurrentHashMap ensures thread safety.
+	 *    
+	 * 3. An ApplicationStateListener. 
+	 *    On startup the ASL only removes applications from the list of apps that have shut down so they will not return a no-op
+	 *    On shutdown the ASL cleans up an OpenTelemetry object, removes its parent OpenTelemetryInfo from the cache, and registers it as an app which has shut down.
+	 */
+
+	private static final String INSTRUMENTATION_NAME = "io.openliberty.microprofile.telemetry";
+	private static final String ENV_DISABLE_PROPERTY = "OTEL_SDK_DISABLED";
+	private static final String CONFIG_DISABLE_PROPERTY = "otel.sdk.disabled";
+	private static final String ENV_METRICS_EXPORTER_PROPERTY = "OTEL_METRICS_EXPORTER";
+	private static final String CONFIG_METRICS_EXPORTER_PROPERTY = "otel.metrics.exporter";
+	private static final String ENV_LOGS_EXPORTER_PROPERTY = "OTEL_LOGS_EXPORTER";
+	private static final String CONFIG_LOGS_EXPORTER_PROPERTY = "otel.logs.exporter";
+	private static final String SERVICE_NAME_PROPERTY = "otel.service.name";
+
+	private static final TraceComponent tc = Tr.register(OpenTelemetryInfoFactory.class);
+
+	private static final Map<J2EEName, OpenTelemetryInfo> appToInfo = new ConcurrentHashMap<J2EEName, OpenTelemetryInfo>();
+	private static final Set<J2EEName> shutDownApplications = Collections.newSetFromMap(new WeakHashMap<J2EEName, Boolean>()); //We have no good place to say forget about an app entirely, so I use weakreferences. 
+
+	//The key methods that provide access to telemetry objects begin here
+	public static OpenTelemetryInfo getOpenTelemetryInfo(J2EEName j2EEName) {
+		return appToInfo.computeIfAbsent(j2EEName, OpenTelemetryInfoFactory::createOpenTelemetryInfo);
+	}
+	
+	public static Tracer getTracer(J2EEName j2EEName) {
+		try {
+			return getOpenTelemetryInfo(j2EEName).getOpenTelemetry().getTracer(INSTRUMENTATION_NAME);
+		} catch (Exception e) {
+			Tr.error(tc, Tr.formatMessage(tc, "CWMOT5002.telemetry.error", e));
+			return OpenTelemetry.noop().getTracerProvider().get("");
+		}
+	}
+
+	public static Span getSpan() {
+		return new SpanProxy(); //TODO I cannot see a way to get no-op spans or baggage, check if we need code here to handle apps that have shut down.
+	}
+
+	public static Baggage getBaggage() {
+		return new BaggageProxy();
+	};
```

Comment:
No-op `Span`: [Span.getInvalid()](https://www.javadoc.io/static/io.opentelemetry/opentelemetry-api/1.29.0/io/opentelemetry/api/trace/Span.html#getInvalid())

There isn't really a "no-op" `Baggage`, but we can get an empty one: [Baggage.empty()](https://www.javadoc.io/static/io.opentelemetry/opentelemetry-api/1.29.0/io/opentelemetry/api/baggage/Baggage.html#empty())

Reply:
That said, I'm not sure it's necessary here. `SpanProxy` and `BaggageProxy` will call `Span.current()` and `Baggage.current()` respectively which should do something sensible.

I'm not sure there's any need for these objects outside CDI though, where we need an object which will reflect the current `Span` or `Baggage` changing. I think most other code could just call `Span.current()` or `Baggage.current()` whenever they need it, so we could omit these methods and create the objects in the Producer.

Reply:
I was thinking its good to make the CDI and static entry points to open-telemetry identical, but if you think calling `current()` here is better I can make the change.